### ISSUE TITLE
Add Safari 10.1 nomodule fix

### ISSUE THREAD

**Issue by web-padawan** [DanielSchaffer/webpack-babel-multi-target-plugin#9]
First of all, thanks for your work! I have implemented the build using `BabelMultiTargetPlugin` and received a feedback about the code broken in Safari 10.1 and iOS 10.3. 

The issue is that Safari 10.1+ does not support `nomodule` attribute (fixed in 11). It is a well known issue mentioned e.g. in [blog post](https://jakearchibald.com/2017/es-modules-in-browsers/) by Jake Archibald. So, the code is getting executed twice

This issue especially breaks Custom Elements registration: 1) they cannot be redeclared, and 2) they use ES2015 classes, and when transpiled to ES5, they need a special script, otherwise the error `TypeError: Constructor requires 'new' operator` will be thrown.

Known workaround: https://gist.github.com/samthor/64b114e4a4f539915a95b91ffd340acc

Here is how it is injected by [ModernModePlugin](https://github.com/vuejs/vue-cli/blob/e7af0d8fa374311693f14d1605e7ae442e43c3a6/packages/%40vue/cli-service/lib/webpack/ModernModePlugin.js#L70) used in Vue CLI.

Please consider adding an option to include this fix, e.g. based on presence of Safari 10.1 and iOS 10.1 - 10.3 in the list of targeted browsers. From my experience, some users still support Safari 10.x.

Original issue reported here: https://github.com/web-padawan/polymer3-webpack-starter/commit/fb82b28a0ae697cdfed3a34e88bd81aa40f059ba#r31858089
I confirmed the issue myself using Simulator: works in iOS 9 (ES5) and broken in 10.3.

**Comment by DanielSchaffer** [comment]
Oh that's no good! I'll try and take a look this week. I'm also totally open to a PR, though I'd prefer to implement the option as part of the options passed to the plugin - trying to check the browserlist seems a bit opaque. Thank you for the report!

**Comment by moderndeveloperllc** [comment]
@DanielSchaffer Just an FYI, iOS 10.3 is still showing [about 5% worldwide](http://gs.statcounter.com/ios-version-market-share/mobile-tablet/worldwide/#monthly-201712-201812) while [Safari 10.1 is less than 1%](http://gs.statcounter.com/browser-version-partially-combined-market-share/desktop/worldwide/#monthly-201712-201812), so this fix is mainly for mobile users.

Edit: Oof. I'm bad at reading charts. iOS 10.3 is 5% of Safari's share which is 10% global, 20% US, so the overall % is rather low.

**Comment by web-padawan** [comment]
Yes the option makes sense, especially note how Vue CLI takes care about `unsafe-inline` CSP rule.

**Comment by DanielSchaffer** [comment]
@moderndeveloperllc @web-padawan Take a look at [v2.1.0-alpha.1](https://github.com/DanielSchaffer/webpack-babel-multi-target-plugin/releases/tag/v2.1.0-alpha.1). This adds a top-level option for `BabelMultiTargetPlugin` that will automatically embed @samthor's workaround into the HTML page generated by `HtmlWebpackPlugin`. I used BrowserStack to confirm that it works in both MacOS Safari 10.1, as well as iOS 10.

So far I've only tested it against the Angular example, as the Angular CLI gives you a basic Protractor setup, so I've still got some work to do to create a standardized "does this load correctly" test that will translate across all the examples.

However, this _should_ at least get you unblocked. I'm totally open to suggestions on the implementation, so feel free to take a peek at #10.

I'm not 100% happy with the fact that it's limited to people using `HtmlWebpackPlugin`, but I figure that people _not_ using that would prefer, if not require, stricter control over what gets injected into their templates, and can handle adding this fix manually if they need it.

**Comment by moderndeveloperllc** [comment]
The only real issue I have is that a base64 `src` means that my CSP policy will need to include `script-src data:` which is just as bad as `script-src unsafe-inline` as it allows arbitrary JS execution on a page. But is it in the purview of a babel webpack plugin like this to include a `safari.nomodule.fix.js` file into the build? I do like that you've made the default to be `true` as this is an issue that some may not know about. Hopefully, this can be dropped in a year or two.

**Comment by DanielSchaffer** [comment]
@moderndeveloperllc I think you might've misread - I haven't put this on by default (at least, not yet) - it's just turned on in the shared configuration for the examples. I'm a little reticent to make this an opt-out type feature, as I realize that injecting a script like this can be a touchy subject. Leaving it off by default would be in line with the rest of the browser-specific workarounds available related tools like Babel and Uglify.

With regards to CSP - I must admit that I'm not terribly familiar with this, but I don't see why the option couldn't be an enum:
* `false` - disables the feature entirely, of course
* `'inline-base64'` - uses the current functionality - I chose that simply because it's more compact than plain text
* `'inline'` - same as above, but plaintext instead of base64
* `'external'` - adds the script to the build outputs, and uses the url for the src instead

**Comment by moderndeveloperllc** [comment]
@DanielSchaffer Sorry, I was looking at the example and not the code. Thinking about it, you are right that having the fix included by default isn't best. Lots of folks couldn't care less about the issue. Having an enum would definitely provide the most flexibility.

A good Content Security Policy is tricky. Google has a couple of concise paragraphs about the [problem with inline JS code](https://developers.google.com/web/fundamentals/security/csp/#inline_code_is_considered_harmful) if you want to learn more. Important stuff if you work in high security sectors with lots of regulations about "best industry practices". Thanks for all your work with this. It really helps provide an alternative to `polymer build`.

**Comment by DanielSchaffer** [comment]
So, hypothetically, if none of the "inline" type options are available due to a CSP, we're looking at either adding another script tag like I said above, or injecting a script into one of the bundles.

I'm not nuts about the extra tag route, because it necessarily adds an extra blocking HTTP request for EVERYONE when it's not even an issue for most users. We couldn't even do a conditional load, because of course that itself would requiring using an inline script.

Injecting it into the bundle seems more streamlined, and a bit more like a polyfill, but that's an entirely different solution from the plugin's perspective as it's no longer just dealing with the `HtmlWebpackPlugin`. It could also be a far more complicated implementation - hopefully, it'll just be as simple as inserting a chunk before all the others, but I'm not 100% positive if that'll be enough to ensure it gets executed before anything else. I'm guessing I'd also have to mess around with using `defer` on the script that doesn't have the workaround injected to ensure the correct timing.

I don't mind doing this of course, just want to get your thoughts from the perspective of someone with CSP requirements before I started off down that road :)

**Comment by web-padawan** [comment]
Thanks for the quick fix! Confirmed to work at web-padawan/polymer3-webpack-starter@ca58d082e2868628992b612c8d864e2a03122f60
The updated [live demo](https://polymer3-webpack-starter.herokuapp.com/employee-new) is now working in my iOS simulator 10.3 so the original issue is fixed.

Regarding the CSP, your idea from the above comment (https://github.com/DanielSchaffer/webpack-babel-multi-target-plugin/issues/9#issuecomment-453702427) sounds good to me.

**Comment by moderndeveloperllc** [comment]
Having an additional HTTP call is a tradeoff that those of us who need to be tight with CSP face quite a bit. If a developer were to use `external`, it's because they had already resigned themselves to making those extra `script` calls. At least HTTP/2 can help some. A note in the documentation would be sufficient warning for the unaware. I also totally understand if you aren't interested in further pursuing an issue that affects so few (tight CSP _and_ Safari 10.1).

I'm not sure putting the fix into a bundle would work necessarily as even the first webpack bundle can come after `nomodule` calls for scripts like the helpers needed to use Babel with ES5 browsers. @web-padawan's project has this in `index.html`.
```
<script nomodule src="./vendor/babel-helpers.min.js"></script>
<script nomodule src="./vendor/regenerator-runtime.min.js"></script>
<script src="./vendor/webcomponents-loader.js"></script>
<script type="module" src="main.c17d38c5.js"></script> //webpack ESM bundle
<script type="text/javascript" src="main.es5.b4d1e632.js" nomodule></script>  //webpack ES5 bundle
```

One way to mitigate the extra call that `external` would introduce is to have the external script itself be `nomodule`:
```
<script nomodule src="safari10NoModuleFix.min.js"></script>
```
This would at least spare about 80% of browsers from loading it. FWIW, I'm not sure you need to include the entire fix comment. Using only the link to the gist and a title would save a few bytes while also maintaining proper attribution.

**Comment by DanielSchaffer** [comment]
regarding the babel helpers and those other polyfills - shouldn't those be automatically included in the legacy bundle? or have you turned that off in your config?

**Comment by DanielSchaffer** [comment]
@web-padawan @moderndeveloperllc [v2.1.0-alpha.2](https://github.com/DanielSchaffer/webpack-babel-multi-target-plugin/releases/tag/v2.1.0-alpha.2) adds an implementation for `safariNoModuleFix: 'external'`

**Comment by DanielSchaffer** [comment]
@web-padawan @moderndeveloperllc not having a "bundled" (read: not requiring another HTTP request) option was bugging me, so I did a bit of playing around. Essentially you were correct in that apply that specific `nomodule` fix is pretty much a non-starter from within one of the bundles due to the timing of when the scripts are loaded. However, I was able to come at it from a completely different angle.

Webpack does a little monkey patching on a global array (`window.webpackJsonp`) - from what I can tell, when chunks are loaded, they get pushed to that array, and webpack replaces the array's `push` function with its own `webpackJsonpCallback` function. By adding a simple check to webpack's bootstrapping logic that checks the name of the array's `push` function, I can just stop execution of a bundle if it detects that the function has already been replaced. Then, this also changes `BabelMultiTargetPlugin`'s logic so that it uses the original module IDs instead of the "tagged" ones (e.g. `@angular/core` instead of `@angular/core?babel-target=modern`), which ensures that for vendor chunks, a given module is only loaded once even if the browser loads both the "modern" and "legacy" chunks.

This serves as an alternative to the `safariNoModuleFix` option, and is enabled by adding the option `normalizeModuleIds: true` to `BabelMultiTargetPlugin`. Since this does seem a little more "hacky", I've marked this feature as "experimental" for now. It's available in [v2.1.0-alpha.3](https://github.com/DanielSchaffer/webpack-babel-multi-target-plugin/releases/tag/v2.1.0-alpha.3)
